sdk/core/azure-core/src/main/java/com/azure/core/util/serializer/JsonCapable.java

```diff
@@ -0,0 +1,32 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+package com.azure.core.util.serializer;
+
+/**
+ * Indicates that the implementing class can be serialized to and deserialized from JSON.
+ * <p>
+ * Since deserialization needs to work without an instance of the class implementing this interface it's assume the
+ * class has a static method {@code fromJson(JsonReader)} that deserializes and instance of that class. The contract for
+ * reading JSON from {@link JsonReader} is that the initial state of the reader on call will either be a null {@link
+ * JsonToken} or be the {@link JsonToken} after the {@link JsonToken#FIELD_NAME} for the object. So, for objects calling
+ * out to other {@link JsonCapable} objects for deserialization, they'll pass the reader pointing to the token after the
+ * {@link JsonToken#FIELD_NAME}. This way objects reading JSON will be self-encapsulated for reading properly formatted
+ * JSON. And, if an error occurs during deserialization an {@link IllegalStateException} should be thrown.
+ *
+ * @param <T> The type of the object that is JSON capable.
+ */
+public interface JsonCapable<T extends JsonCapable<T>> {
```

Comment:
Should this be called `JsonSerializable` ?

Reply:
I think that's' a good name too, for now I'm going to leave this as-is given it being a working state. Once this is being formalized we can dig further into naming, where code lives, etc.